### PR TITLE
v1.0.1 (errata) 

### PR DESCRIPTION
**Pull request by tetron**
Criteria to be considered for v1.0.1: must not affect syntax of v1.0 documents.  Existing v1.0 documents must remain valid.  May affect behavior for documents that previously relied on undefined or underspecified behavior.

Needs tests:

- [x] Test for location of files in InitialWorkDir
- [x] Test for recursively writable: https://github.com/common-workflow-language/common-workflow-language/pull/382
- [x] Test of workflow in compound document
- [x] Test that `default` applies when source is `null`
- [x] Test that missing parameters are assigned `null` by default
- [x] Conformance test for nameroot and nameext #454

Documentation todos:

- [x] Merge discussion from #348 
- [x] Merge updated schema-salad documentation to include documentation of mapping transforms and type DSL
- [x] Scatter over empty list is empty
- [x] Clarify interpretation of secondaryFiles on inputs (all files are expected to be present, no optionals)
- [x] What to do if "format" field on output is an array: #482 
- [ ] clarify the subject of every imperative sentences: for each sentence, is this the user's job or the implementation's job? @mr-c

Additional release checklist:

- [ ] Get sign off from SBG, other stakeholders
- [ ] Update docs
  - [x] Add a detailed "What's changed" section for `v1.0.1`

Post-merge:

- [ ] Add github release
- [ ] Upload to figshare, get updated DOI

### CODE REVIEW

**Comment by tetron** [comment]
Needs tests:

[ ] Test for location of files in InitialWorkDir
[*] Test for recursively writable: https://github.com/common-workflow-language/common-workflow-language/pull/382
[ ] Test that for slashes as separators in ids and anchor ids that start with `#`

**Comment by tetron** [comment]
Todo: merge discussion from #348

**Comment by mr-c** [comment]
Probably should be v1.0.14 or later: https://github.com/common-workflow-language/common-workflow-language/commit/cd50eb7a06b89caa3aff80220fa7d6054071e7b3

**Comment by tetron** [comment]
> Probably should be v1.0.14 or later: cd50eb7

None of those changes were actually bundled into a formal release, though.  Unless we're going to go back and make packages and get DOIs for each of those point releases, I think it would be easier (and less confusing) to consolidate everything under a formal v1.0.1 release that gets the full release treatment.  All of those changes should certainly be in the changelog for v1.0.1.

**Comment by tetron** [comment]
Additional release checklist:

- [ ] Get sign off from SBG, other stakeholders
- [ ] Update docs

Post-merge:

- [ ] Add github release
- [ ] Upload to figshare, get updated DOI

@ntijanic @StarvingMarvin

**Comment by tetron** [comment]
@geoffjentry

> 3.5 of command line spec has a typo: EMACScript

**Comment by tetron** [comment]
Consider adding LoadListingRequirement to v1.0.1 since it is an extension and not a syntax change.

File: v1.0/CommandLineTool.yml [hunk 1]
```diff
@@ -787,6 +796,13 @@ $graph:
         May be an expression.  If so, the expression return value must validate
         as `{type: array, items: [File, Directory]}`.
 
+        Files or Directories which are listed in the input parameters and
+        appear in the `InitialWorkDirRequirement` listing must have their
+        `path` set to their staged location in the designated output directory.
```
Comment (StarvingMarvin) [inline v1.0/CommandLineTool.yml]: I'm fine with this, but this is in contradiction with #396 where Directory in InitialWorkdirRequirement lacks path (which is also a reason why bunny fails that test).

**Comment by StarvingMarvin** [comment]
One thing that should also be clarified is behavior of the scattering over an empty list, especially the cases with multiple scatters.

**Comment by mr-c** [comment]
Re: the DOI & Figshare. I uploaded `v1.0` right over `draft-3` as Figshare has version control and appends a `vN` to the DOI as new versions are uploaded.

https://doi.org/10.6084/m9.figshare.3115156.v1 -> `draft-3`
https://doi.org/10.6084/m9.figshare.3115156.v2 -> `v1.0`
which will make
https://doi.org/10.6084/m9.figshare.3115156.v3 -> `v1.0.1`

**Comment by tetron** [comment]
Another corner case is the proper interpretation of `secondaryFiles` on inputs.  In this example:

```
    - id: vcfNames
      type: File[]
      inputBinding:
          position: 5
          prefix: --vcfs
      secondaryFiles:
          - ".tbi"
          - ".idx"
```

Does this mean (a) all listed secondaryFiles are expected (b) at least one secondaryFile is expected (c) exactly one secondaryFile is expected (d) is purely advisory and must not be used for input validation.  For reference, the current behavior of `cwltool` is (a) on the assumption that a primary file might have multiple secondary files, but interpretation (b) would support the case where there are several possible valid extensions for a secondaryFile.

**Comment by mr-c** [comment]
I nominate #482 for inclusion

**Comment by mr-c** [comment]
re: optional secondaryFiles: not now

**Comment by mr-c** [comment]
FYI: https://github.com/common-workflow-language/common-workflow-language/pull/454/

**Comment by mr-c** [comment]
@tetron will add test case using compound (`--packed`) document

File: v1.0/Workflow.yml [hunk 2]
```diff
@@ -66,6 +66,20 @@ $graph:
...
+        * 12 March 2017:
+          * Mark `default` as not required for link checking.
+          * Add note that recursive subworkflows is not allowed.
+          * Fix mistake in discussion of extracting field names from workflow step ids.
+        * 21 July 2017:
+          * Add clarification about scattering over empty arrays.
+          * Clarify interpretation of secondaryFiles on inputs.
+        * 22 July 2017: (v1.0.1)
+          * Expanded discussion of semantics of File and Directory types
+
```
Comment (mr-c) [inline v1.0/Workflow.yml]: other non mentioned changes
- EMACScript -> E**CMA**Script
- Clarified processing of input object's input parameters when a value is not provided
- Clarified valid types and meaning of the format field on inputs versus outputs
- in the context of a `valueFrom` Expression: Input parameter defaults must be applied before evaluating the
expression

**Comment by mr-c** [comment]
- [ ] workflow.yml missing new version, line 13

**Comment by tetron** [comment]
Updated release notes @mr-c

**Comment by mr-c** [comment]
How about https://github.com/common-workflow-language/common-workflow-language/pull/403 ?

**Pull request merged**